### PR TITLE
Speed up DeepTau by running the inner/outer convolution in batch

### PR DESCRIPTION
**Pull request by hqucms**
#### PR description:

Running the inner/outer convolution in batch for each tau candidate speeds up the overall DeepTau inference by ~20%.

Related issue: https://github.com/cms-sw/cmssw/issues/27119

#### PR validation:

Validated on a JetHT file with ~100k events and obtained consistent results on DeepTau outputs.

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@hqucms 
do you have a profiler report for this PR improvements?
If you do, please post a link.
Thank you.

**Comment by hqucms** [comment]
@slava77 

From the timing report:
```
TimeReport   0.054420     0.054420     0.054420  deepTau2017v2p1   (baseline)
TimeReport   0.044977     0.044977     0.044977  deepTau2017v2p1   (w/ this PR)
```

And IgProf reports:
 - baseline: http://hqu.web.cern.ch/hqu/dev/cgi-bin/igprof-navigator/NANO-JetHT2017D-CMSSW_11_0_0_pre10-TIMING/67
 - w/ this PR: http://hqu.web.cern.ch/hqu/dev/cgi-bin/igprof-navigator/NANO-JetHT2017D-CMSSW_11_0_0_pre10-PR28128-DeepTauBatching-TIMING/72

All produced with the standard NanoAOD workflow in `CMSSW_11_0_0_pre10` (which has DeepTau enabled), using an input from the JetHT 2017D sample.

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc [hunk 1]
```diff
@@ -1114,6 +1116,42 @@ class DeepTauId : public deep_tau::DeepTauBase {
                           const CellGrid& grid,
                           bool is_inner) {
     tensorflow::Tensor& convTensor = *convTensor_.at(is_inner);
+
+    eGammaTensor_[is_inner] = std::make_unique<tensorflow::Tensor>(
+        tensorflow::DT_FLOAT,
+        tensorflow::TensorShape{
+            (long long int)grid.num_valid_cells(), 1, 1, dnn_inputs_2017_v2::EgammaBlockInputs::NumberOfInputs});
```
Comment (slava77) [inline RecoTauTag/RecoTau/plugins/DeepTauId.cc]: this approach increases the size of the tensors in memory.
My very rough calculations add up to about 250 kB extra if the grids are fully filled.
This seems acceptable.
If this is correct, no action is needed.

Comment (hqucms) [reply RecoTauTag/RecoTau/plugins/DeepTauId.cc]: That's roughly what I get as well. 
In practice, the grids should be quite sparse (<10% filled based on https://indico.cern.ch/event/829640/contributions/3473229/attachments/1869221/3097000/2019-06-26_DeepTau_ID_computational_performance.pdf).

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc [hunk 2]
```diff
@@ -1081,7 +1083,7 @@ class DeepTauId : public deep_tau::DeepTauBase {
     }
   }
 
-  void getPartialPredictions(tensorflow::Tensor& convTensor, bool is_inner, int eta_index, int phi_index) {
+  tensorflow::Tensor getPartialPredictions(unsigned batch_size, bool is_inner) {
```
[the remark this answers is not in the document]
Comment (hqucms) [reply RecoTauTag/RecoTau/plugins/DeepTauId.cc]: You are right, it's not needed here. Looks like I forgot to clean it up when submitting the PR. It's fixed in https://github.com/cms-sw/cmssw/pull/28128/commits/5977afeccc314d709551c319ff0284429b67c069.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/28128	5977afe 
- code changes are in line with the PR description and the follow up review
- jenkins tests pass and comparisons with the baseline show no differences
    - technical performance improvements were provided in https://github.com/cms-sw/cmssw/pull/28128#issuecomment-543167886

**Comment by fabiocos** [comment]
+1

**Pull request merged**